java-manta-client-unshaded/src/main/java/com/joyent/manta/client/MantaBucketListingIterator.java

```diff
@@ -0,0 +1,485 @@
+     */
+    private final int pagingSize;
+
+    /**
+     * Prefix filter that helps in optimizing a buckets listing.
+     */
+    private volatile String prefix = null;
+
+    /**
+     * Filter to group names with a common prefix ending in its first occurrence.
+     */
+    private Character delimiter = null;
+
+    /**
+     * Path to bucket in which we will iterate through its contents.
+     */
+    private final String path;
+
+    /**
+     * HTTP request helper class.
+     */
+    private final HttpHelper httpHelper;
+
+    /**
+     * The total number of lines that we have iterated through.
+     */
+    private final AtomicLong lines = new AtomicLong(0);
+
+    /**
+     * The next line of data that we haven't iterated to yet.
+     */
+    private final AtomicReference<String> nextLine = new AtomicReference<>();
+
+    /**
+     * Flag indicated if we have finished and there is nothing left to iterate.
+     */
+    private final AtomicBoolean finished = new AtomicBoolean(false);
+
+    /**
+     * Jackson JSON parsing instance.
+     */
+    private final ObjectMapper mapper = MantaObjectMapper.INSTANCE;
+
+    /**
+     * The marker we use to request against the Manta API.
+     */
+    private volatile String marker;
+
+    /**
+     * The current {@link BufferedReader} instance that wraps the HTTP response
+     * {@link java.io.InputStream} from our most recent request to the API.
+     */
+    private volatile BufferedReader br;
+
+    /**
+     * The most recent response object from the page of data that we are currently
+     * parsing.
+     */
+    private volatile CloseableHttpResponse currentResponse;
+
+    /**
+     * Create a new instance of a bucket list iterator.
+     *
+     * @param path path to bucket in which we will iterate through its contents
+     * @param httpHelper HTTP request helper class
+     * @param prefix prefix filter that helps in optimizing a buckets listing
+     * @param pagingSize size of result set requested against the Manta API (2-1024)
+     * @param delimiter filter to group names with a common prefix ending in its first occurrence
+     */
+    public MantaBucketListingIterator(final String path,
+                                      final HttpHelper httpHelper,
+                                      final String prefix,
+                                      final Character delimiter,
+                                      final int pagingSize) {
+        Validate.notBlank(path, "Path must not be blank");
+        Validate.notNull(httpHelper, "HTTP help must not be null");
+        Validate.notNull(prefix, "Prefix parameter must not be null");
+        Validate.notNull(delimiter, "Delimiter parameter must not be null");
+
+        this.path = path;
+        this.httpHelper = httpHelper;
+
+        if (!(pagingSize >= 2 && pagingSize <= MAX_RESULTS)) {
+            throw new IllegalArgumentException("Paging size must be greater than "
```

Comment:
nit: the logic is `pagingSize >= 2` but the warning is `> 1`... even though they are effectively the same thing, i'd recommend changing the wording or the code to align them.